Fix `!= <array>` for unfiltered nulls and missing array of nulls

jeeminso opened the pull request:
## Summary of the changes / Why this improves CrateDB
Fixes:
```
cr> show create table t;
+-----------------------------------------------------+
| SHOW CREATE TABLE doc.t                             |
+-----------------------------------------------------+
| CREATE TABLE IF NOT EXISTS "doc"."t" (              |
|    "a" ARRAY(INTEGER)                               |
| )                                                   |
| CLUSTERED INTO 4 SHARDS                             |

cr> select * from t;
+--------+
| a      |
+--------+
| [null] |
| NULL   |
| [1]    |
| []     |
+--------+

cr> select * from t where a != [];
+-----+
| a   |
+-----+
| [1] | -- missing [null]
+-----+
```


```
cr> show create table t5;
+-----------------------------------------------------+
| SHOW CREATE TABLE doc.t5                            |
+-----------------------------------------------------+
| CREATE TABLE IF NOT EXISTS "doc"."t5" (             |
|    "a" ARRAY(INTEGER) STORAGE WITH (                |
|       columnstore = false                           |
|    )                                                |
| )                                                   |
| CLUSTERED INTO 4 SHARDS                             |

cr> select * from t5;
+--------+
| a      |
+--------+
| []     |
| NULL   |
| [1]    |
| [null] |
+--------+

cr> select * from t5 where a != [1];
+--------+
| a      |
+--------+
| []     |
| NULL   | -- unfiltered
| [null] |
+--------+
```

We did not handle for the cases where refExistsQuery was null:
https://github.com/crate/crate/pull/16427/files#diff-b98ffbe4dd33f483188fef958602ca8513268138164548a489a98b25e2da00bcR236-R241

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

jeeminso commented:
```
# Results (server side duration in ms)
V1: 5.9.0-8c9f118b7166284c8889acb4e040a59300ebffe4
V2: 5.9.0-38347f72a48f010623526a3123f92842b2e3b259

Q: select count(*) from t_arrays_biased_empty where xs != []
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       25.032 ±  231.358 |      2.754 |      3.578 |      3.966 |   3159.030 |
|   V2    |      268.149 ±  386.383 |    201.878 |    211.517 |    214.170 |   4176.083 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               + 165.85%                           + 193.35%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 165.85%
The test has statistical significance

Q: select count(*) from t_arrays_biased_one_element where xs != []
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       16.336 ±  104.263 |      5.547 |      9.365 |      9.655 |   1482.800 |
|   V2    |       50.111 ±  142.842 |     23.828 |     34.934 |     41.355 |   1532.998 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               + 101.66%                           + 115.44%   
There is a 99.28% probability that the observed difference is not random, and the best estimate of that difference is 101.66%
The test has statistical significance
```

jeeminso commented on server/src/main/java/io/crate/expression/predicate/NotPredicate.java:
```diff
@@ -229,10 +229,16 @@ public Query toQuery(Function input, LuceneQueryBuilder.Context context) {
-                    countEmptyArrays(context.parentQuery(), nullableRef, context));
+                    context
+                );
                 if (refExistsQuery != null) {
                     builder.add(refExistsQuery, BooleanClause.Occur.MUST);
+                } else {
+                    // fall back
+                    return new BooleanQuery.Builder()
+                        .add(notX, Occur.MUST)
+                        .add(LuceneQueryBuilder.genericFunctionFilter(input, context), Occur.FILTER)
+                        .build();
```
Basically we did not handle for the cases where `refExistsQuery` was null.

jeeminso commented:
retest this please - unrelated failures

jeeminso commented on server/src/main/java/io/crate/expression/predicate/IsNullPredicate.java:
```diff
@@ -109,32 +109,28 @@ public Query toQuery(Function function, Context context) {
 
     @Nullable
-    public static Query refExistsQuery(Reference ref, Context context, boolean countEmptyArrays) {
+    public static Query refExistsQuery(Reference ref, Context context) {
```
`countEmptyArrays` can be removed because when `ref` is an array type, `countEmptyArrays` passed in is always `true`.

BaurzhanSakhariev replied:
Do I get it right, that we needed these parameter because it used to be called in array_length and in https://github.com/crate/crate/pull/16329 we changed `ArrayUpperFunction` to not use it anymore?

jeeminso replied:
Yes, that were 3 call sites that called `refExistsQuery` with `countEmptyArrays` as `false`, `ArrayUpperFunction` and `AnyNeqOperator`. The first two were removed by #16329 and the other is irrelevant since `ref` is never an array type as the method name suggests, `AnyNeqOperator.refMatchesAnyArrayLiteral`(`ref != ANY(<array_literal>)`).

BaurzhanSakhariev commented on server/src/main/java/io/crate/expression/predicate/NotPredicate.java:
```diff
@@ -229,10 +229,16 @@ public Query toQuery(Function input, LuceneQueryBuilder.Context context) {
                     builder.add(refExistsQuery, BooleanClause.Occur.MUST);
+                } else {
+                    // fall back
+                    return new BooleanQuery.Builder()
+                        .add(notX, Occur.MUST)
```
afaik, we add MUST before if/else a initial state of the builder
` builder.add(notX, BooleanClause.Occur.MUST);`, so we end up adding it twice.

I think in fallback we should add only `.add(LuceneQueryBuilder.genericFunctionFilter(input, context), Occur.FILTER `

jeeminso replied:
I thought we could fall back to https://github.com/crate/crate/blob/6d1907556c74f45af55a674598b6ed3b4c04b6ec/server/src/main/java/io/crate/expression/predicate/NotPredicate.java#L220-L223
instead of continuiting to build on top of https://github.com/crate/crate/blob/6d1907556c74f45af55a674598b6ed3b4c04b6ec/server/src/main/java/io/crate/expression/predicate/NotPredicate.java#L225-L226

Could you double check?

BaurzhanSakhariev replied:
ah right, we have a new Builder here, sorry

BaurzhanSakhariev reviewed (commented):
LGTM, thanks!

Added a question and a minor comment.

Not approving as it would be great to do a follow up https://github.com/crate/crate/pull/16427#issuecomment-2257020457 here as a separate commit for better visibility

jeeminso commented:
- status update:
If I recall correctly, I need https://github.com/crate/crate/pull/16436/commits/3abbd2dc5e58aca8754a32d8317da698d8e35e48#diff-ac8d1ac2fedf242f9df1adebdd342edf18e25ad626fa3c2e697e2efe8766cf71R43-R63 merged in order to handle the test failure.

BaurzhanSakhariev commented:
Is this up to date?

>f I recall correctly, I need https://github.com/crate/crate/commit/3abbd2dc5e58aca8754a32d8317da698d8e35e48#diff-ac8d1ac2fedf242f9df1adebdd342edf18e25ad626fa3c2e697e2efe8766cf71R43-R63 merged in order to handle the test failure.

I guess it's done and it's ready for review?

BaurzhanSakhariev commented on docs/appendices/release-notes/5.9.0.rst:
```diff
@@ -59,6 +59,15 @@ Breaking Changes
   regressions on tables created before 5.9.0. For tables created on and after
   5.9.0, the fix has positive impact on the performance.
 
+- Fixed an issue that caused ``WHERE`` clause containing ``NOT`` operator on
+  an array type against an empty array to incorrectly filter array of nulls. It
+  is a breaking change because the fix causes performance degradations.
+
+- Fixed an issue that caused ``WHERE`` clause containing ``NOT`` operator on
+  an array type without doc-values against a non-empty array to incorrectly
+  un-filter null rows. It is a breaking change because the fix causes
+  performance degradations.
+
```
2 questions:

1. Do I get it right, that this PR also fixes https://github.com/crate/crate/issues/16476? 
I guess we don't want to backport because it's a breaking change and not a bug fix?

2. Shouldn't we phrase it similar to https://github.com/crate/crate/commit/6c63ebf5feee7ef8db1a6a3158e416d98053166d
ie add
  ...```
regressions on tables created before 5.9.0. For tables created on and after
  5.9.0, the fix has positive impact on the performance
```

to both entries?

Also, could you please measure the impact of the second commit, ie compare it with numbers you [posted](https://github.com/crate/crate/pull/16427#issuecomment-2257020457) for the first commit?

jeeminso replied:
> Do I get it right, that this PR also fixes https://github.com/crate/crate/issues/16476?
I guess we don't want to backport because it's a breaking change and not a bug fix?

Yes, we should not backport this.

> Shouldn't we phrase it similar to https://github.com/crate/crate/commit/6c63ebf5feee7ef8db1a6a3158e416d98053166d

Thanks, added

BaurzhanSakhariev reviewed (approved):
LGTM, thanks!

Added some questions

jeeminso commented:
Just noticed `NumTermsPerDocQuery` was still used for filtering by `= []`(which is a special case and the tests added here did not catch it). Will merge https://github.com/crate/crate/pull/16479 before this.

jeeminso commented:
Addressed remaining comments and here are the benchmark results with array index queries:

```
# Results (server side duration in ms)
V1: 5.9.0-db39420cf38352beb9495045482c974a6207bd89
V2: 5.9.0-499ff8b765a425f39e105544d810e4a613082427

Q: select count(*) from t_arrays_biased_empty where xs != []
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        6.919 ±   13.138 |      3.324 |      5.563 |      6.225 |    195.782 |
|   V2    |       10.944 ±    6.958 |      8.330 |     10.224 |     11.154 |    128.370 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +  45.06%                           +  59.04%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 45.06%
The test has statistical significance

Q: select count(*) from t_arrays_biased_one_element where xs != []
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |        4.055 ±   12.418 |      1.705 |      2.512 |      2.842 |    113.072 |
|   V2    |        4.107 ±    4.305 |      1.599 |      2.195 |      2.744 |     16.214 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               +   1.27%                           -  13.50%   
There is a 5.46% probability that the observed difference is not random, and the best estimate of that difference is 1.27%
The test has no statistical significance
```

jeeminso commented:
In addition to https://github.com/crate/crate/pull/16427#issuecomment-2313772853, we can measure the performance of the second commit with a more direct query, `<array> is null`:
```
# Results (server side duration in ms)
V1: 5.9.0-db39420cf38352beb9495045482c974a6207bd89
V2: 5.9.0-499ff8b765a425f39e105544d810e4a613082427

Q: select count(*) from t where a is null -- a is an array of int column with ~4M rows of null and [1]
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |      361.490 ±  898.700 |    251.674 |    260.037 |    268.513 |  10422.155 |
|   V2    |        1.658 ±    4.412 |      0.796 |      1.251 |      1.534 |     77.311 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               - 198.17%                           - 198.09%   
There is a 100.00% probability that the observed difference is not random, and the best estimate of that difference is 198.17%
The test has statistical significance
```
